src/zkbob/manager/MPCGuard.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+import "../../../src/zkbob/ZkBobPool.sol";
+import "../../utils/Ownable.sol";
+import "../utils/CustomABIDecoder.sol";
+
+import "../../interfaces/IZkBobPool.sol";
+
+contract MPCGuard is Ownable, CustomABIDecoder {
+    address[] private guards;
+
+    address operator;
+
+    address public immutable pool;
+
+    uint256 constant SIGNATURE_SIZE = 64;
+
+    constructor(address _operator, address _pool) Ownable() {
+        pool = _pool;
+        _setOperator(_operator);
+    }
+
+    /**
+     * @dev Throws if called by any account other than the current relayer operator.
+     */
+    modifier onlyOperator() {
+        require(operator == _msgSender(), "ZkBobPool: not an operator");
+        _;
+    }
+
+    function _setOperator(address _operator) internal {
+        operator = _operator;
+    }
+
+    function setOperator(address _operator) external onlyOwner {
+        _setOperator(_operator);
+    }
+
+    function setGuards(address[] calldata _guards) external onlyOwner {
+        guards = _guards;
+    }
+
+    function digest(bytes memory data) internal pure returns (bytes32) {
+        return ECDSA.toEthSignedMessageHash(keccak256(data));
+    }
+
+    modifier calldataVerified() {
+        (uint8 count, bytes calldata signatures) = _mpc_signatures();
+        require(count == guards.length, "MPCWrapper: wrong quorum");
+        ZkBobPool poolContract = ZkBobPool(pool);
+        uint256 currentRoot = poolContract.roots(poolContract.pool_index());
```

Comment:
Master operator will use the latest root from the optimistic state for guard signature requests, right?

Reply:
Yes:  by the time this transaction request will hit the chain, this optimistic root will either become finalized or whole queue will be purged and reprocessed.